### PR TITLE
Introduce `Dep`

### PR DESCRIPTION
**Pull request by tindzk**

### CODE REVIEW

**Comment by mkotsbak** [comment]
What is Dep supposed to be used for, comparing to the other data structures?

File: shared/src/main/scala/pl/metastack/metarx/Dep.scala [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  private val internal = Sub[U](fwd(sub))
+  internal.attach(u => value = u)
+
+  sub := sub.get
+
+  def :=(channel: ReadChannel[U]): Unit = sub := bwd(channel)
+  def :=(value: U): Unit = sub := bwd(Var(value))
+  def get: U = value
+
+  def toReadChannel: ReadChannel[U] = fwd(sub)
```
Comment (darkfrog26) [inline shared/src/main/scala/pl/metastack/metarx/Dep.scala]: I wasn't sure if there way any way to accomplish the same result, but make `Dep` actually extend `ReadChannel[U]`?

File: shared/src/main/scala/pl/metastack/metarx/Dep.scala [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+/** @see [[Sub.dep()]] */
+private[metarx] class Dep[T, U](sub: Sub[T],
+                                fwd: ReadChannel[T] => ReadChannel[U],
+                                bwd: ReadChannel[U] => ReadChannel[T]) {
+  private var value: U = _
+
+  private val internal = Sub[U](fwd(sub))
+  internal.attach(u => value = u)
+
+  sub := sub.get
```
Comment (darkfrog26) [inline shared/src/main/scala/pl/metastack/metarx/Dep.scala]: Is there a better way to accomplish set update of `value` without having to do this?

File: shared/src/main/scala/pl/metastack/metarx/Dep.scala [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package pl.metastack.metarx
+
+/** @see [[Sub.dep()]] */
+private[metarx] class Dep[T, U](sub: Sub[T],
+                                fwd: ReadChannel[T] => ReadChannel[U],
+                                bwd: ReadChannel[U] => ReadChannel[T]) {
+  private var value: U = _
```
Comment (darkfrog26) [inline shared/src/main/scala/pl/metastack/metarx/Dep.scala]: Technically, do we need this?  Can't we actually just have `def get: U = internal.get`?

Comment (tindzk) [inline shared/src/main/scala/pl/metastack/metarx/Dep.scala]: Indeed. Do you want to remove `value`?

File: shared/src/main/scala/pl/metastack/metarx/Dep.scala (hunk 2 above)
Comment (tindzk) [inline shared/src/main/scala/pl/metastack/metarx/Dep.scala]: Why is it necessary in the first place?

File: shared/src/main/scala/pl/metastack/metarx/Dep.scala (hunk 1 above)
Comment (tindzk) [inline shared/src/main/scala/pl/metastack/metarx/Dep.scala]: That may work, although you can also inherit from `Dep`.

File: shared/src/main/scala/pl/metastack/metarx/Dep.scala [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+                                fwd: ReadChannel[T] => ReadChannel[U],
+                                bwd: ReadChannel[U] => ReadChannel[T]) {
+  private val channel = fwd(sub)
+  private val internal = Sub[U](channel)
+
+  def :=(channel: ReadChannel[U]): Unit = sub := bwd(channel)
+  def :=(value: U): Unit = sub := bwd(Var(value))
+  def get: U = internal.get
+
+  def toReadChannel: ReadChannel[U] = channel
```
Comment (darkfrog26) [inline shared/src/main/scala/pl/metastack/metarx/Dep.scala]: I'm still hopeful we'll be able to make `Dep` itself a `ReadChannel`.  Is the only way to accomplish this to extend `ReadChannel[U]` and override every method to call `channel`?

Comment (tindzk) [inline shared/src/main/scala/pl/metastack/metarx/Dep.scala]: You could try to inherit from `Sub[U]`.

**Comment by tindzk** [comment]
@mkotsbak Consider `right = x + width`. Obviously the following holds: `x = right - width`. This pull request allows you to express such dependencies in code:

```scala
val right = x.dep[Double](_ + width, _ - width)
```

If you update `right`, it will re-compute `x`. This is not only limited to equations, but can be used for any invertible operation.

In the future, we might consider shortening the syntax because for most mathematical operators we already know the inverse operation:

```scala
val right = (x + width).dep(x)
```

**Comment by mkotsbak** [comment]
Hmm, but how is it different to Channel.bimap?

**Comment by darkfrog26** [comment]
I'm okay with this being merged if you are.

**Comment by darkfrog26** [comment]
BTW, should it eventually be:

```scala
val right = x.dep(_ + width)
```

It seems like that would be a lot cleaner than:

```scala
val right = (x + width).dep(x)
```

**Comment by tindzk** [comment]
@mkotsbak The difference is that `biMap` only deals with single values, whereas `Dep` is an extension of `Sub` which supports channel subscriptions. For example, have a look at line 65 in [DepSpec](https://github.com/MetaStack-pl/MetaRx/blob/master/shared/src/test/scala/pl/metastack/metarx/DepSpec.scala#L65).

If you have any ideas how to make the API more consistent, please let us know!

**Pull request merged**